Powder Elastic mode for the DNS Reduction GUI

koshchii opened the pull request:
**Description of work.**

This PR integrates an elastic powder data reduction workflow into the DNS Reduction GUI.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

- Download test data: [Elastic_Powder.zip](https://github.com/mantidproject/mantid/files/11040197/Elastic_Powder.zip)
- Open the reduction GUI [Interfaces->Direct->DNS Reduction]
- Browse to the directory where the test data are located using the "Paths" tab of the GUI
- Load and select the data for reduction using the "Data" tab
- Choose the data reduction options using the "Options" tab
- Reduce the data and generate a data reduction script using the "Script Generator" tab
- Visualize results using the "Plotting" tab 

<!-- Instructions for testing. -->

Fixes: Part 2 of #27538. <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

thomueller committed:
first commit of elastic powder GUI

koshchii committed:
Add minor style changes

koshchii committed:
Add minor style changes

koshchii committed:
Fix the problem with generating output Nexus files

koshchii committed:
Remove unused methods in main_view; fix typo

koshchii committed:
Add powder elastic tests to CmakeLists

koshchii committed:
Fix failing unit tests

koshchii committed:
Add documentation file

koshchii committed:
Refactor fselector -> f_selector

koshchii committed:
Implement auto title on GUI's opening

koshchii committed:
Fix script generation problem for standard data

koshchii committed:
Remove unimplemented 'Filter Standard Data for Measured 2θ' corr opt

koshchii committed:
Remove standard interpolation option and related files

koshchii committed:
Remove unnecessary create_script_method in dns_dataset

koshchii committed:
Update Options UI

koshchii committed:
Refactor tests

koshchii committed:
Fix dns_dataset.py not to try filling in empty dicts

koshchii committed:
Fix broken binning in dns_dataset.py designed for interpolated data

koshchii committed:
Minor update of the plotting tab

koshchii committed:
Implement renaming of dns_dataset to dns_powder_elastic_dataset

koshchii committed:
Fix failing tests

koshchii committed:
Update elastic pwd dataset filereader to work similar to tof pwd

koshchii committed:
Refactor + implement error messages in script generator model

koshchii committed:
Fix broken binning

koshchii committed:
Implement removal of not needed std data bank angles & autobin funct

koshchii committed:
Remove not needed fields parameter from DNSElasticDataset class

koshchii committed:
Fix binning bugs, remove nans in MDHistoWS for integration purposes

koshchii committed:
Implement limits on two theta step when mannual binning is chosen

koshchii committed:
dns_powder_tof_maintenance_updates

koshchii committed:
Implement refactoring and add minor style changes

koshchii committed:
Upload documentation

koshchii committed:
Update documentation

koshchii committed:
Refact and clean up generated workspaces in md_powder_elastic

koshchii committed:
Fix xyz_separation and implement respective tailored error message

koshchii committed:
Implement flipping_ratio_correction check; refactor xyz_data

koshchii committed:
Fix coherent/incoherent separation check

koshchii committed:
Fix destructive masking of MDHistoWorkspaces, add warning on masking

koshchii committed:
Modify default plotting style in Plotting tab

koshchii committed:
Fix automatic_omega_binning in dns_elastic_powder_dataset

koshchii committed:
Modify error messages in script_generator_model

koshchii committed:
Update UI for plotting tab

koshchii committed:
Update Options view, presenter, and add binning

koshchii committed:
Separate dim binning and binning in md_powder_elastic

koshchii committed:
Implement removal of selected std data not used for reduction

koshchii committed:
Remove redundant nicr std data from generated script

koshchii committed:
Refactor get_elastic_standard_data_dic

koshchii committed:
Fix broken Options UI

koshchii committed:
Fix failing tests

koshchii committed:
Update helpers_for_testing

koshchii committed:
Merge branch 'mantidproject:main' into dns_gui_powder_elastic_dev

koshchii committed:
Fix pre-commit errors

koshchii committed:
Fix pre-commit formatting errors

koshchii committed:
Add release note

gemmaguest commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_elastic/command_line/command_check.py:
```diff
@@ -0,0 +1,72 @@
+import os
```
Missing copyright

gemmaguest commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_elastic/data_structures/dns_binning.py:
```diff
@@ -0,0 +1,26 @@
+import numpy as np
```
Missing copyright

gemmaguest commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_elastic/plot/elastic_powder_plot_model.py:
```diff
@@ -0,0 +1,110 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+
```
Some of the copyright statements have a blank line breaking them up. Probably worth fixing because I know in the past we've had automated tools that modify the copyright statements, so it's good if they all match the same format.

gemmaguest commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_powder_tof/main_view.py:
```diff
@@ -43,10 +43,14 @@ def __init__(self, parent=None, app=None, within_mantid=None):
         self.modus_mapping = {
+            self.ui.actionPowder_Elastic: "powder_elastic",
             self.ui.actionPowder_TOF: "powder_tof",
+            # self.ui.actionSingle_crystal_elastic: 'sc_elastic',
```
Are these commented out lines supposed to be here?

gemmaguest commented on qt/python/mantidqtinterfaces/test/dns_powder_elastic/command_line/command_check_test.py:
```diff
@@ -0,0 +1,96 @@
+import unittest
```
Missing copyright

gemmaguest commented on qt/python/mantidqtinterfaces/test/dns_powder_elastic/plot/elastic_powder_plot_presenter_test.py:
```diff
@@ -0,0 +1,163 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+
```
Blank line

gemmaguest commented on qt/python/mantidqtinterfaces/test/dns_powder_elastic/plot/elastic_powder_plot_model_test.py:
```diff
@@ -0,0 +1,153 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+
```
Blank line

gemmaguest commented on qt/python/mantidqtinterfaces/test/dns_powder_elastic/data_structures/field_names_test.py:
```diff
@@ -0,0 +1,23 @@
+# Mantid Repository : https://github.com/mantidproject/mantid
+#
+# Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+
```
Blank line

gemmaguest reviewed (changes requested):
I haven't done a thorough review, but the code quality looks good, it follows a nice MVP design and test coverage looks good. Good work :+1:

Unfortunately it crashes when I generate the script! It seems to get most way through processing so is perhaps when it completes? I am testing on Ubuntu 22.04

I've also added a few minor comments around formatting of copyright statements, and also the release notes will need moving.

koshchii committed:
Implement changes from code review

koshchii committed:
Merge branch 'mantidproject:main' into dns_gui_powder_elastic_dev

koshchii committed:
Merge remote-tracking branch 'refs/remotes/origin/dns_gui_powder_elastic_dev' into dns_gui_powder_elastic_dev

koshchii committed:
Remove comments in main_view of powder_tof

koshchii committed:
Fix crash on exec of ConvertMDHistoToMatrixWorkspace and formatting

koshchii committed:
Fix failing tests

sf1919 commented:
Given the age of this PR you may need to rebase it because of the move to Python 3.10. We've noticed failing tests from some PRs are a result of this rather than an issue with the code.

koshchii committed:
Merge branch 'mantidproject:main' into dns_gui_powder_elastic_dev

koshchii commented:
> I haven't done a thorough review, but the code quality looks good, it follows a nice MVP design and test coverage looks good. Good work 👍
> 
> Unfortunately it crashes when I generate the script! It seems to get most way through processing so is perhaps when it completes? I am testing on Ubuntu 22.04
> 
> I've also added a few minor comments around formatting of copyright statements, and also the release notes will need moving.

Thanks much for the review! Now, the script generator should be able to complete reduction without crashing. The crashing happened due to the default setting of the parameter 'FindXAxis' of ConvertMDHistoToMatrixWorkspace. Once I added 'FindXAxis'=False (non-default value) to ConvertMDHistoToMatrixWorkspace in the script generator, reduction finished successfully. The rest of the suggestions are also implemented now.

koshchii commented:
> Given the age of this PR you may need to rebase it because of the move to Python 3.10. We've noticed failing tests from some PRs are a result of this rather than an issue with the code.

Should I open another PR from another branch that is rebased on the top of the main branch and close this PR? As far as I understand, if I force push the rebased changes to this branch, the history will be rewritten. Or I am misunderstanding something?

sf1919 commented:
I think given today is code freeze and this PR has passed all it's CI tests then I will leave it for @gemmaguest to finish reviewing rather than suggest we rebase.

gemmaguest commented:
Thanks for the changes, it all looks good and now generates the script without crashing. The script seems to run ok, produces lots of output workspace and saves them to ascii. The outputs seem to have lots of nan's in them and I don't see anything on the plot, but this is likely me running it with the wrong settings so I am not concerned about this if the developer is happy with it. 

The only other thing is that the embedded plot has a weird resolution. This might be my setup (running Ubuntu over a remote connection) so again I am not too worried about this. 

![image](https://github.com/mantidproject/mantid/assets/12895056/743308df-e09a-4b2e-91db-cbccef1d441b)

Overall, I don't see anything risky here so despite the large size of the PR I am happy for this to go into the release.

gemmaguest commented:
@sf1919 The release notes have been copied to v6.8 but not removed from v6.7. Just wondering if we should remove them or if they will just be ignored?

sf1919 commented:
This can be ignored as it will be picked up by the release editors.

koshchii committed:
Remove release note from v6.7.0

koshchii commented:
Sorry about forgetting to remove the release note from v6.7.0. Now it is only in 6.8. Regarding the white screen, I remember having this issue in the past and I addressed it with replacing nans (and inf) with 0 (the user gets the corresponding warning message) for the most common selection of combinations of sample/standard data (and corresponding angles, fields, binning, etc). However, there are way too many various possible combinations of sample and standard data (and reduction choices) that a user could choose, so one needs to see if the chosen selection in "Data" and "Options" tabs is feasible and if yes, it needs to be addressed.

gemmaguest reviewed (approved):
Thanks for fixing the release notes and thanks Tom for fixing the base branch. This looks good to me now :+1:
